src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+    }
+
+    const prevVersionId = await this.prev(versionId);
+    if (prevVersionId === null) {
+      return null;
+    }
+
+    return this.first(prevVersionId);
+  }
+
+  /**
+   * Return the next version of a DID document if it exists or null, otherwise.
+   */
+  public next (versionId: VersionId): VersionId | null {
+    const nextVersionId = this.nextVersion.get(versionId);
+    if (nextVersionId === undefined) {
+      return null;
+    } else {
+      return nextVersionId;
+    }
+  }
+
+  /**
+   * Return the latest (most recent) version of a DID document. Return null if
+   * the version is unknown.
+   */
+  public last (versionId: VersionId): VersionId | null {
+    const opInfo = this.opHashToInfo.get(versionId);
+    if (opInfo === undefined) {
+      return null;
+    }
+
+    const nextVersionId = this.next(versionId);
+    if (nextVersionId === null) {
+      return versionId;
+    } else {
+      return this.last(nextVersionId);
+    }
+  }
+
+  /**
+   * Get the last processed transaction.
+   * TODO: fix this after discussing the intended semantics.
+   */
+  public get lastProcessedTransaction (): Transaction | undefined {
+    return undefined;
+  }
+
+  /**
+   * Get a cryptographic hash of the write operation. Currently, uses
+   * SHA256 to get hashes (TODO: Fix it to be consistent DID generation)
+   */
+  private static getHash (operation: WriteOperation): OperationHash {
+    const sha256HashCode = 18;
+    const multihash = Multihash.hash(operation.operationBuffer, sha256HashCode);
+    const multihashBase58 = Base58.encode(multihash);
+    return multihashBase58;
+  }
+
+  /**
+   * Apply state changes for operations that have a previous version (update, delete, recover)
+   */
+  private applyOpWithPrev (opHash: OperationHash, opInfo: OperationInfo, version: VersionId): void {
+    // We might already know of an update to this version. If so, we retain
+    // the older of previously known update and the current one
+    const prevUpdateHash = this.nextVersion.get(version);
+    if (prevUpdateHash !== undefined) {
+      const prevUpdateInfo = this.opHashToInfo.get(prevUpdateHash) as OperationInfo;
+      if (lesser(prevUpdateInfo, opInfo)) {
+        return;
+      }
+    }
+
+    this.nextVersion.set(version, opHash);
+  }
+
+  /**
+   * Return true if the provided operation is an initial version i.e.,
+   * produced by a create operation.
+   */
+  private isInitialVersion (opInfo: OperationInfo): boolean {
+    return opInfo.type === OperationType.Create;
+  }
 
   /**
-   * Applies the given DID operation to the DID Cache.
+   * Return the operation given its (access) info.
    */
-  apply (operation: WriteOperation): void;
+  private async getOperation (opInfo: OperationInfo): Promise<WriteOperation> {
+    const batchBuffer = await this.cas.read(opInfo.batchFileHash);
+    const batch = JSON.parse(batchBuffer.toString());
+    const opBuffer = Buffer.from(batch[opInfo.operationIndex].data);
```

Comment:
Batch file not following protocol specified schema.

Reply:
You are correct - but I think this is consistent with current rooter code (that was what I used as a reference) - please check.

Reply:
I've fixed this.